### PR TITLE
[Workflow] Add a `TraceableWorkflow`

### PR DESCRIPTION
**Pull request by lyrixx**
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       |
| License       | MIT
| Doc PR        |

image

image

---

Initially, I wanted to display the current marking on the graph, but it's not possible. "current" does not mean anything. What if I use the workflow for 2 entities? What if I call twice `apply()`. Which marking should I choose? => this is not possible.

But, DoctrineBundle use ajax, like for explaining SQL request. It would be nice to be able to click on a marking, and it'll re-render the graph with this marking. WDYT?

### CODE REVIEW

File: src/Symfony/Component/Workflow/TraceableWorkflow.php [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    public function getCalls(): array
+    {
+        return $this->calls;
+    }
+
+    private function callInner(string $method, array $args): mixed
+    {
+        $startAt = microtime(true);
```
Comment (94noni) [inline src/Symfony/Component/Workflow/TraceableWorkflow.php]: why not using a debug.stopwatch? to have the performance tab "for free"

File: src/Symfony/Component/Workflow/TraceableWorkflow.php [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Workflow;
```
Comment (94noni) [inline src/Symfony/Component/Workflow/TraceableWorkflow.php]:
```suggestion
namespace Symfony\Component\Workflow\Debug;
```

i am used to seeing some debug classes inside a debug subnamespace

File: src/Symfony/Component/Workflow/Debug/TraceableWorkflow.php [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        try {
+            return $return = $this->workflow->{$method}(...$args);
+        } catch (\Throwable $exception) {
+            throw $exception;
+        } finally {
+            $periods = $this->stopwatch->stop($sMethod)->getPeriods();
+            $period = end($periods);
+            $call = [
+                'method' => $method,
+                'duration' => $period->getDuration(),
```
Comment (94noni) [inline src/Symfony/Component/Workflow/Debug/TraceableWorkflow.php]: thx for having swap to stopwatch, what bout adding here:

```suggestion
                'duration' => $period->getDuration(),
                'memory' => $period->getMemory(),
```

would be awesome, as per internal usage we can plug some ±heavy listener/subscriber on workflow event, and having both duration and memory can help fast check on what cant be improved :)

Comment (lyrixx) [reply src/Symfony/Component/Workflow/Debug/TraceableWorkflow.php]: I'm not sure it's really useful, since here it's not per event, but for the whole method (`can`, `apply`, etc). So you won't know what use memory. 

I'm working on adding profiler for each listener - but to be honest, it's super hard :(

Comment (94noni) [reply src/Symfony/Component/Workflow/Debug/TraceableWorkflow.php]: copy that, lets see how it goes :)

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/WorkflowDebugPass.php [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace Symfony\Bundle\FrameworkBundle\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+use Symfony\Component\Workflow\Debug\TraceableWorkflow;
+
+/**
+ * Adds all configured security voters to the access decision manager.
```
Comment (94noni) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/WorkflowDebugPass.php]: comment seems wrong

**Review by 94noni — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/WorkflowDebugPass.php (hunk 4 above)
Comment (lyrixx) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/WorkflowDebugPass.php]: thanks, fixed

File: src/Symfony/Component/Workflow/Debug/TraceableWorkflow.php (hunk 3 above)
Comment (stof) [reply src/Symfony/Component/Workflow/Debug/TraceableWorkflow.php]: Profiling memort is also quite hard. As soon as you have nested periods, things become wonky because the memory used by the inner traceable object now counts as well for the outside span.
Memory profiling is better done with an actual profiler tool than this traceable instrumentation. It will be less skewed.

Comment (HeahDude) [reply src/Symfony/Component/Workflow/Debug/TraceableWorkflow.php]: >I'm working on adding profiler for each listener - but to be honest, it's super hard :(

@lyrixx It should be done by default if the workflow uses the event dispatcher service, it is traceable in debug and events should be listed in the Events profiler panel since always.

Comment (lyrixx) [reply src/Symfony/Component/Workflow/Debug/TraceableWorkflow.php]: @HeahDude indeed, the profiler shows some data, but it's not as good as what you can see in the Security profiler

### Workflow

image

### Security 

image


### TL;DR:

I would like something with

* all listener called
  * timing information
* for the guard
  * the "response"

**Comment by lyrixx** [comment]
This PR is ready for 2 month. I'm gonna merge it by the end of the week if there are not more review

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/WorkflowDebugPass.php [hunk 5]
```diff
@@ -0,0 +1,37 @@
+<?php
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/WorkflowDebugPass.php]: This class should probably be part of the component.

**Comment by fabpot** [comment]
@lyrixx Last call for 6.4 :)

File: src/Symfony/Component/Workflow/DataCollector/WorkflowDataCollector.php [hunk 6]
```diff
@@ -57,4 +68,36 @@ public function getWorkflows(): array
...
+            TransitionBlocker::class => function ($v, array $a, Stub $s, $isNested) {
+                unset(
+                    $a[sprintf(Caster::PATTERN_PRIVATE, $v::class, 'code')],
+                    $a[sprintf(Caster::PATTERN_PRIVATE, $v::class, 'parameters')],
+                );
+
+                return $a;
+            },
+            Marking::class => function ($v, array $a, Stub $s, $isNested) {
+                $a[sprintf(Caster::PATTERN_PRIVATE, $v::class, 'places')] = array_keys($v->getPlaces());
```
Comment (stof) [inline src/Symfony/Component/Workflow/DataCollector/WorkflowDataCollector.php]: shouldn't this use the virtual prefix instead ?

File: src/Symfony/Component/Workflow/DataCollector/WorkflowDataCollector.php [hunk 7]
```diff
@@ -57,4 +68,36 @@ public function getWorkflows(): array
...
+    }
+
+    protected function getCasters()
+    {
+        $casters = [
+            ...parent::getCasters(),
+            TransitionBlocker::class => function ($v, array $a, Stub $s, $isNested) {
+                unset(
+                    $a[sprintf(Caster::PATTERN_PRIVATE, $v::class, 'code')],
+                    $a[sprintf(Caster::PATTERN_PRIVATE, $v::class, 'parameters')],
```
Comment (stof) [inline src/Symfony/Component/Workflow/DataCollector/WorkflowDataCollector.php:89]: Shouldn't this mark the stub as cut ?

File: src/Symfony/Component/Workflow/Debug/TraceableWorkflow.php [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+            try {
+                $previousMarking = $this->workflow->getMarking($args[0]);
+            } catch (\Throwable) {
+            }
+        }
+
+        try {
+            return $return = $this->workflow->{$method}(...$args);
+        } catch (\Throwable $exception) {
+            throw $exception;
```
Comment (stof) [inline src/Symfony/Component/Workflow/Debug/TraceableWorkflow.php]: This catch block should be removed entirely. It makes no sense.

Comment (stof) [reply src/Symfony/Component/Workflow/Debug/TraceableWorkflow.php]: OK, actually, it is not useless because it assigns the variable for the finally clause.
Maybe the code should be refactored without using finally but separate registration of the failed and successful calls.

File: src/Symfony/Component/Workflow/Debug/TraceableWorkflow.php (hunk 3 above)
Comment (stof) [inline src/Symfony/Component/Workflow/Debug/TraceableWorkflow.php]: Shouldn't this use the event duration instead of only the last period ?

File: src/Symfony/Component/Workflow/Registry.php [hunk 9]
```diff
@@ -43,7 +43,10 @@ public function has(object $subject, string $workflowName = null): bool
         return false;
     }
 
-    public function get(object $subject, string $workflowName = null): Workflow
+    /**
+     * @return WorkflowInterface
+     */
+    public function get(object $subject, string $workflowName = null)
```
Comment (stof) [inline src/Symfony/Component/Workflow/Registry.php]: If there was already a native return type, why widening it to `mixed` to restrict it again in the next major version instead of widening it to WorkflowInterface ?

**Comment by lyrixx** [comment]
Hello @fabpot, @stof and others reviewer. 

I addressed your comments, I think the PR is ready 👍🏼

File: src/Symfony/Component/Workflow/DataCollector/WorkflowDataCollector.php [hunk 10]
```diff
@@ -57,4 +68,38 @@ public function getWorkflows(): array
...
+                );
+
+                $s->cut += 2;
+
+                return $a;
+            },
+            Marking::class => function ($v, array $a, Stub $s, $isNested) {
+                $a[Caster::PREFIX_VIRTUAL] = array_keys($v->getPlaces());
```
Comment (stof) [inline src/Symfony/Component/Workflow/DataCollector/WorkflowDataCollector.php]: you should still name your property after the prefix.

```suggestion
                $a[Caster::PREFIX_VIRTUAL.'places'] = array_keys($v->getPlaces());
```

Comment (lyrixx) [reply src/Symfony/Component/Workflow/DataCollector/WorkflowDataCollector.php]: rah, 🤦🏼 thanks !

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @lyrixx.
